src/librustc_mir/interpret/eval_context.rs

```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
         FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), Pointer<M::PointerTag>>,
 }
 
+struct SpanGuard(log::span::Span);
+
+impl SpanGuard {
+    fn new() -> Self {
+        Self(log::span::Span::none())
+    }
+
+    fn enter(&mut self, span: log::span::Span) {
+        if let Some(id) = span.id() {
+            log::dispatcher::get_default(|d| {
+                d.enter(&id);
+            });
```

Comment:
This is a slightly different behavior than what `tracing`'s `Span` type does, and (potentially) subtly incorrect if multiple subscribers are in use. 

The `Span` struct [tracks the dispatcher that observed the span's creation](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing/src/span.rs#L363-L367) (the call to `new_span`) and gave it an `Id`. Subsequently, when the span is entered, exited, or dropped, it [ensures that those methods are called on the original subscriber](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing/src/span.rs#L716-L718), even if it has moved to the context of a _different_ subscriber. 

This is because span ID assignment is only guaranteed to be unique on a per-subscriber basis. A different subscriber may have assigned the same ID to an unrelated span. If multiple subscribers are in use, this could result in a messed-up trace. 

I think we could implement the same behavior as `tracing::Span` here if we used [`Span::with_subscriber`](https://docs.rs/tracing/0.1.18/tracing/span/struct.Span.html#method.with_subscriber), like this:
```suggestion
        span.with_subscriber(|(id, dispatch)| {
            dispatch.enter(id);
        });
```

Reply:
ooh, that makes sense.